src/codegen/expression.rs

```diff
@@ -2425,7 +2425,13 @@ fn array_subscript(
 
     let mut array = expression(array, cfg, contract_no, func, ns, vartab, opt);
     let index_ty = index.ty();
+    // Array indexes must be integers, so if the index is a reference to a type, we must dereference it.
     let index = expression(index, cfg, contract_no, func, ns, vartab, opt);
+    let index = if let Type::Ref(ty) = index.ty() {
+        Expression::Load(index.loc(), *ty, Box::new(index))
+    } else {
+        index
+    };
```

Comment:
The way we usually do this is using `let index = index.cast(loc, index_ty.deref_any(), true, ns, diagnostics)?;`. This also solved the case where `index` is storage variable and need loading from storage.

Reply:
> This also solved the case where `index` is storage variable and need loading from storage.

This is not true.

Reply:
When I refactored codegen last year, I removed loads from the cast function.

Reply:
The `Expression::Load` should be generated in sema.

Reply:
This problem is very similar to https://github.com/hyperledger/solang/pull/1112

Reply:
> The `Expression::Load` should be generated in sema.

This is what I've done now. IMO, loads should be done in codegen, but that requires a large refactoring and is something for another PR.